rust/dyn-ddlog/src/ast/tree.rs

```diff
@@ -0,0 +1,651 @@
+use crate::ast::rewrites::{BlockSimplifier, ExpressionRewriter, NestingSimplifier};
+use std::{
+    fmt::{self, Debug, Write},
+    slice, vec,
+};
+
+pub use crate::resolve::Symbol;
+
+// TODO: Source locations
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct Declaration<I> {
+    pub attributes: Vec<Attribute<I>>,
+    pub kind: DeclarationKind<I>,
+}
+
+impl<I> Declaration<I> {
+    pub const fn new(attributes: Vec<Attribute<I>>, kind: DeclarationKind<I>) -> Self {
+        Self { attributes, kind }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub enum DeclarationKind<I> {
+    Relation(Relation<I>),
+    Rule(Rule<I>),
+    Fact(Fact<I>),
+    Function(Function<I>),
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct Relation<I> {
+    pub kind: RelationKind,
+    pub name: I,
```

Comment:
Should this be `Symbol` or `string` rather than `I` (i.e., this is just a name, not a resolved symbol, correct?). Same goes for some other  declarations.

Reply:
Symbols can be traded resolved into their source strings, keeping things as symbols is very convenient